AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1282,19 +1281,51 @@ - (void)_tearDownPlaceholderLayer
 
 #pragma mark - For Subclasses
 
+- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  if (_methodOverrides & ASDisplayNodeMethodOverrideCalculateSizeThatFits) {
+    CGSize size = [self calculateSizeThatFits:constrainedSize.max];
+    return [ASLayout newWithLayoutableObject:self size:ASSizeRangeClamp(constrainedSize, size)];
+  } else {
+    id<ASLayoutable> layoutSpec = [self layoutSpecThatFits:constrainedSize];
+    ASLayout *layout = [layoutSpec measureWithSizeRange:constrainedSize];
+    // Make sure layoutableObject of the root layout is `self`, so that the flattened layout will be structurally correct.
+    if (layout.layoutableObject != self) {
+      layout.position = CGPointZero;
```

Comment:
This condition is interesting to me.  Do we expect it to happen without programmer error?  Can we add a basic assertion in the condition if we don't expect it to happen, but subclasses might accidentally trigger it from improper implementation?  In this case we still want to keep the fallback behavior because such conditions might only be encountered for the production usage of an app.